Reverse proxy

mpgermano opened the pull request:
Implemented Reverse Proxy Handler
Updated Response and Request objects
Fixed find_prefix bug

Todo:
  Reverse Proxy Unit Tests
  Reverse Proxy Integration Test

mpgermano commented:
The Travis script is failing due to a bug in the find_prefix function of the webserver. I'll try to fix this.

schung507 commented:
Were you able to test 302 redirect? The /about/chancellor page is a 302 redirect on the ucla page but it doesnt show up when i run the server.

mpgermano commented:
@schung507 Yeah just looked, there must be a bug somewhere. I'll get it figured out.

tyhchoi commented:
Hey I looked into the find_prefix bug and I don't think you need to change anything. In the config file you set root to be /home/mpgermano/CS130/Please-Hire-Us/www, but the first slash is not necessary. localhost:2020/ also redirects to the ucla page in the original implementation as well.

mpgermano commented:
@tyhchoi You're right. Thanks for point this out. I'll revert those changes

mpgermano commented:
Fixed the /about/chancellor bug. We reverted the changes for find_prefix but paths like /js/plugins/example.js weren't being routed to the correct handler as the "longest" variable was being returned as an empty string. Thus, the small change I made is hacky but works for now.

llam15 commented on src/Webserver.cc:
```diff
@@ -154,10 +154,9 @@ void Webserver::session(tcp::socket sock) {
             // Read the request.
             size_t request_length = sock.read_some(boost::asio::buffer(request), error);
 
-            printf("Connected to client.\n");
```
How come this log was removed?

llam15 commented on src/Webserver.cc:
```diff
@@ -154,10 +154,9 @@ void Webserver::session(tcp::socket sock) {
-
             const std::unique_ptr<Request> req = Request::Parse(std::string(request, request_length));
-            RequestHandler* handler = get_handler(req->uri());
+            
```
Can you fix the spacing here?

llam15 commented on src/reverse_proxy_handler.cc:
```diff
@@ -0,0 +1,206 @@
+    for (int i = 0; i < MAX_REDIRECTS; i++){
+      std::cout << "Beginning to forward request in ReverseProxyHandler to " << host_ << std::endl;
+      
+      Response::ResponseCode forwardRC = ForwardRequest(transformedRequest, response, nextHost);
```
I find it a little weird that ForwardRequest returns a Response Code that does not correspond with the actual response code, which you check later. Would it be more natural to return a bool indicating whether or not the request was forwarded properly (since ForwardRequest returns either OK or Internal Server Error). Another option would be to return a RequestHandler Status instead (either OK or Proxy Error). Or you could return the actual response code, and then check if Internal Server Error was returned.

llam15 commented:
Also, the integration test for the proxy handler is failing? I'm not exactly sure why, but it seems that the diff is producing two different files
